core/state_transition.go

```diff
@@ -551,6 +587,16 @@ func (st *StateTransition) creditTxFees() error {
 	// No need to do effectiveTip calculation, because st.gasPrice == effectiveGasPrice, and effectiveTip = effectiveGasPrice - baseTxFee
 	tipTxFee := new(big.Int).Sub(totalTxFee, baseTxFee)
 
+	if st.msg.MaxFeeInFeeCurrency() != nil {
+		// Celo Denominated
+		tipTxFee = feeCurrencyRate.FromBase(tipTxFee)
+		baseTxFee = feeCurrencyRate.FromBase(baseTxFee)
+		totalTxFee = feeCurrencyRate.FromBase(totalTxFee)
+		refund.Sub(st.erc20FeeDebited, totalTxFee)
+		// No need to exchange gateway fee since it's it's deprecated on G fork,
+		// and MaxFeeInFeeCurrency can only be present in H fork (which implies G fork)
+	}
```

Comment:
We want to ensure that
```
st.erc20FeeDebited = tipTxFee + baseTxFee + refund
```
so that debit and credit totals match. Since the exchange rate conversions have limited accuracy, the only way to achieve this is to calculate one of the three credit values based on the two others *after* the exchange rate conversion.

Example how the current code fails:
```
rateNum = 1
rateDen = 2
tipTxFee = 3
baseTxFee = 3
totalTxFee = 6
st.erc20FeeDebited = 6

tipTxFee = 3 * rateNum // rateDen = 1
baseTxFee = 3 * rateNum // rateDen = 1
totalTxFee = 6 * rateNum // rateDen = 3
refund = st.erc20FeeDebited - totalTxFee = 6 - 3 = 3
assert st.erc20FeeDebited = tipTxFee + baseTxFee + refund  # WRONG! 6 = 1 + 1 + 3
```

```suggestion
	if st.msg.MaxFeeInFeeCurrency() != nil {
		// Celo Denominated
		tipTxFee = feeCurrencyRate.FromBase(tipTxFee)
		baseTxFee = feeCurrencyRate.FromBase(baseTxFee)
		totalTxFee.Add(tipTxFee, baseTxFee)
		refund.Sub(st.erc20FeeDebited, totalTxFee)
		// No need to exchange gateway fee since it's it's deprecated on G fork,
		// and MaxFeeInFeeCurrency can only be present in H fork (which implies G fork)
	}
```

That will change the end of the example to
```
totalTxFee = tipTxFee + baseTxFee = 2
refund = st.erc20FeeDebited - totalTxFee = 6 - 2 = 4
assert st.erc20FeeDebited = tipTxFee + baseTxFee + refund  # Ok, 6 = 1 + 1 + 4
```

Reply:
forgot to reply to this one. I applied your change and used your explanation as comment for anyone who ventures deep into this code and wonders the why.